### ISSUE TITLE
Error in Autoupdateservice

### ISSUE THREAD

**Issue by FinnStutzenstein** [OpenSlides/openslides-autoupdate-service#111]
Request for the current example data:
```
{"collection":"meeting","ids":[1],"fields":{"name":null,"description":null,"start_time":null,"end_time":null,"location":null,"conference_show":null,"conference_auto_connect":null,"conference_los_restriction":null,"conference_stream_url":null,"url_name":null,"is_template":null,"enable_anonymous":null,"export_csv_separator":null,"export_csv_encoding":null,"export_pdf_pagenumber_alignment":null,"export_pdf_fontsize":null,"export_pdf_pagesize":null,"agenda_show_subtitles":null,"agenda_enable_numbering":null,"agenda_number_prefix":null,"agenda_numeral_system":null,"agenda_item_creation":null,"agenda_new_items_default_visibility":null,"agenda_show_internal_items_on_projector":null,"list_of_speakers_amount_last_on_projector":null,"list_of_speakers_amount_next_on_projector":null,"projector_countdown_warning_time":null,"projector_default_countdown_time":null,"list_of_speakers_couple_countdown":null,"list_of_speakers_show_amount_of_speakers_on_slide":null,"list_of_speakers_present_users_only":null,"list_of_speakers_show_first_contribution":null,"motions_default_workflow_id":null,"motions_default_amendment_workflow_id":null,"motions_default_statute_amendment_workflow_id":null,"motions_preamble":null,"motions_default_line_numbering":null,"motions_line_length":null,"motions_reason_required":null,"motions_enable_text_on_projector":null,"motions_enable_reason_on_projector":null,"motions_enable_recommendation_on_projector":null,"motions_show_referring_motions":null,"motions_enable_sidebox_on_projector":null,"motions_show_sequential_number":null,"motions_recommendations_by":null,"motions_statute_recommendations_by":null,"motions_recommendation_text_mode":null,"motions_default_sorting":null,"motions_number_type":null,"motions_number_min_digits":null,"motions_number_with_blank":null,"motions_statutes_enabled":null,"motions_amendments_enabled":null,"motions_amendments_in_main_list":null,"motions_amendments_of_amendments":null,"motions_amendments_prefix":null,"motions_amendments_text_mode":null,"motions_amendments_multiple_paragraphs":null,"motions_supporters_min_amount":null,"motions_supporters_enable_autoremove":null,"motion_poll_default_type":null,"motion_poll_default_100_percent_base":null,"motion_poll_default_majority_method":null,"motion_poll_default_group_ids":null,"motion_poll_ballot_paper_selection":null,"motion_poll_ballot_paper_number":null,"motions_export_title":null,"motions_export_preamble":null,"motions_export_submitter_recommendation":null,"motions_export_follow_recommendation":null,"assignment_poll_default_method":null,"assignment_poll_default_type":null,"assignment_poll_default_100_percent_base":null,"assignment_poll_default_majority_method":null,"assignment_poll_default_group_ids":null,"assignment_poll_add_candidates_to_list_of_speakers":null,"assignment_poll_sort_poll_result_by_votes":null,"assignment_poll_ballot_paper_selection":null,"assignment_poll_ballot_paper_number":null,"assignments_export_title":null,"assignments_export_preamble":null,"users_sort_by":null,"users_enable_presence_view":null,"users_allow_self_set_present":null,"users_enable_vote_weight":null,"users_pdf_welcometitle":null,"users_pdf_welcometext":null,"users_pdf_url":null,"users_pdf_wlan_ssid":null,"users_pdf_wlan_password":null,"users_pdf_wlan_encryption":null,"users_email_sender":null,"users_email_replyto":null,"users_email_subject":null,"users_email_body":null,"custom_translations":null,"id":null,"default_group_id":{"type":"relation","collection":"group","fields":{"name":null,"id":null}},"superadmin_group_id":{"type":"relation","collection":"group","fields":{"name":null,"id":null}},"logo_$_id":{"type":"template"},"font_$_id":{"type":"template"}}}
```

Message in the client
```
zone-evergreen.js:2845 POST https://localhost:8000/system/autoupdate net::ERR_HTTP2_SERVER_REFUSED_STREAM
```

Stacktrace:
```
autoupdate_1        | 2020/10/22 05:30:34 http2: panic serving 172.20.0.12:57112: runtime error: invalid memory address or nil pointer dereference
autoupdate_1        | goroutine 240 [running]:
autoupdate_1        | net/http.(*http2serverConn).runHandler.func1(0xc0002183d0, 0xc000523f8e, 0xc000197500)
autoupdate_1        |   /usr/local/go/src/net/http/h2_bundle.go:5711 +0x16f
autoupdate_1        | panic(0x723840, 0x998e40)
autoupdate_1        |   /usr/local/go/src/runtime/panic.go:969 +0x175
autoupdate_1        | github.com/openslides/openslides-autoupdate-service/internal/restrict.(*templateField).Check(0xc000098ee0, 0x1, 0xc0003738c0, 0x13, 0xc0000b03b0, 0xe, 0x10, 0x0, 0x203000, 0x10, ...)
autoupdate_1        |   /root/internal/restrict/checker.go:124 +0x304
autoupdate_1        | github.com/openslides/openslides-autoupdate-service/internal/restrict.(*Restricter).Restrict(0xc0000a97a0, 0x1, 0xc0004d09f0, 0x13, 0xc00054c390)
autoupdate_1        |   /root/internal/restrict/restrict.go:59 +0x44f
autoupdate_1        | github.com/openslides/openslides-autoupdate-service/internal/autoupdate.(*Autoupdate).RestrictedData(0xc00009b230, 0x7ec5c0, 0xc000266340, 0x1, 0xc000266440, 0x4, 0x4, 0x4, 0x8, 0x203000)
autoupdate_1        |   /root/internal/autoupdate/autoupdate.go:102 +0x209
autoupdate_1        | github.com/openslides/openslides-autoupdate-service/internal/keysbuilder.(*Builder).Update(0xc000060230, 0x7ec5c0, 0xc000266340, 0x0, 0x0)
autoupdate_1        |   /root/internal/keysbuilder/keysbuilder.go:79 +0x4e4
autoupdate_1        | github.com/openslides/openslides-autoupdate-service/internal/keysbuilder.newBuilder(0x7ec5c0, 0xc000266340, 0x7e5fc0, 0xc00009b230, 0x1, 0xc00051a0c0, 0x1, 0x4, 0xc00037b445, 0x3, ...)
autoupdate_1        |   /root/internal/keysbuilder/keysbuilder.go:34 +0xb4
autoupdate_1        | github.com/openslides/openslides-autoupdate-service/internal/keysbuilder.ManyFromJSON(0x7ec5c0, 0xc000266340, 0x7f31bc16c4d8, 0xc0004d07b0, 0x7e5fc0, 0xc00009b230, 0x1, 0x10, 0xc0001b0aa0, 0x7f31bc4714b0)
autoupdate_1        |   /root/internal/keysbuilder/json.go:66 +0x73d
autoupdate_1        | github.com/openslides/openslides-autoupdate-service/internal/http.(*Handler).complex(0xc0000a97c0, 0xc0004a2600, 0x1, 0x0, 0x0, 0x0, 0x0)
autoupdate_1        |   /root/internal/http/http.go:90 +0xfe
autoupdate_1        | github.com/openslides/openslides-autoupdate-service/internal/http.(*Handler).autoupdate.func1(0x7ebc40, 0xc0002183d0, 0xc0004a2600, 0x0, 0x0)
autoupdate_1        |   /root/internal/http/http.go:56 +0x288
autoupdate_1        | github.com/openslides/openslides-autoupdate-service/internal/http.errHandleFunc.ServeHTTP(0xc0000a97e0, 0x7ebc40, 0xc0002183d0, 0xc0004a2600)
autoupdate_1        |   /root/internal/http/http.go:116 +0x62
autoupdate_1        | github.com/openslides/openslides-autoupdate-service/internal/http.validRequest.func1(0x7ebc40, 0xc0002183d0, 0xc0004a2600)
autoupdate_1        |   /root/internal/http/http.go:190 +0x74
autoupdate_1        | net/http.HandlerFunc.ServeHTTP(0xc0000a9800, 0x7ebc40, 0xc0002183d0, 0xc0004a2600)
autoupdate_1        |   /usr/local/go/src/net/http/server.go:2042 +0x44
autoupdate_1        | net/http.(*ServeMux).ServeHTTP(0xc0000ce440, 0x7ebc40, 0xc0002183d0, 0xc0004a2600)
autoupdate_1        |   /usr/local/go/src/net/http/server.go:2417 +0x1ad
autoupdate_1        | github.com/openslides/openslides-autoupdate-service/internal/http.(*Handler).ServeHTTP(0xc0000a97c0, 0x7ebc40, 0xc0002183d0, 0xc0004a2600)
autoupdate_1        |   /root/internal/http/http.go:39 +0x4d
autoupdate_1        | net/http.serverHandler.ServeHTTP(0xc000170000, 0x7ebc40, 0xc0002183d0, 0xc0004a2600)
autoupdate_1        |   /usr/local/go/src/net/http/server.go:2843 +0xa3
autoupdate_1        | net/http.initALPNRequest.ServeHTTP(0x7ec680, 0xc000183e90, 0xc000147c00, 0xc000170000, 0x7ebc40, 0xc0002183d0, 0xc0004a2600)
autoupdate_1        |   /usr/local/go/src/net/http/server.go:3415 +0x8d
autoupdate_1        | net/http.(*http2serverConn).runHandler(0xc000197500, 0xc0002183d0, 0xc0004a2600, 0xc0004b7f20)
autoupdate_1        |   /usr/local/go/src/net/http/h2_bundle.go:5718 +0x8b
autoupdate_1        | created by net/http.(*http2serverConn).processHeaders
autoupdate_1        |   /usr/local/go/src/net/http/h2_bundle.go:5452 +0x505
```

Please do more testing - also with #109 I have the feeling, that crucial parts of the service are not tested; a not working normal relations must be catched by an test. It looks like, there not enough tests, if even this basic parts are not catched.

**Comment by ostcar** [comment]
You are right. Both issues are located at a part of the code, where there are no tests at all. The reason is, that this part handels the concrete data from the models.yml. All code that is abstract is well tested. But this code is not abstract. To test it, I would need to know how the the models.yml file looks like.

Maybe we could talk via telephone and you could give me some tips how to test it.

**Comment by FinnStutzenstein** [comment]
I would say, that you build your own models.yml where all cases are included. You are indipendent of the real one and can add relations/collections if you need a new test. I can recommend you doing system tests. They come in two fashions:
- A (full) system test: Start your service and all required other ones (redis, datastore) in a docker-compose setup, fill in your own data and have a stand-alone tool, that tests against the autoupdate service, e.g. requesting data, than modify it in the datastore, and check, if the right update is received. This is the most accurate testing, since all other services will use the autoupdate service this way.
- A system test inside your application: You also setup the datastore and redis, but the tests do not go against the autoupdate service. You write tests the same as you are doing it now, but without any mocks. To test everything, your tests are likely to have a single entrypoint - the method/function where a request is received. So you are just leaving the http2-boundary out, but test everything behind it.

The second approach is done by the backend and datastore; the first one by the datastore, too.

**Comment by FinnStutzenstein** [comment]
Another issue for the same request:
```
Internal Error: get first time restricted data: restrict data: checker for key meeting/1/logo_$web_header_id: decoding meeting/1/logo_$web_header_id=3: json: cannot unmarshal number into Go value of type []int
```

**Comment by FinnStutzenstein** [comment]
Maybe you can test it with the client in conjunction. Check out the client to https://github.com/FinnStutzenstein/openslides-client/tree/structuredFields and do a `make run-dev` in the main repo. You will get the errors directly, when the client loads.

**Comment by ostcar** [comment]
I have an Idea how to restructure the code to make it more testable.  I will do it in the next days.

**Comment by ostcar** [comment]
I removed the dependency to the real models.yml data. Now I could write tests with some simple fictive models.yml files. All bugs you found are now fixed and tested. I hope this is the end of this story.

**Comment by FinnStutzenstein** [comment]
Tested with commit `d6ed1fae32c166932abbd9d0d95b9c1624db1f18`. The error `Internal Error: get first time restricted data: restrict data: checker for key meeting/1/logo_$web_header_id: decoding meeting/1/logo_$web_header_id=3: json: cannot unmarshal number into Go value of type []int` remains. Same request as in the first comment in this issue.

**Comment by ostcar** [comment]
After I fixed and tested everything, I forgot to recreate the autogenerated file from the modes.yml... Fixed in #121 

Can you test this again?
